src/plugins/fileextractor/fileextractor.cpp

```diff
@@ -0,0 +1,1634 @@
+            handle);
+
+    if (!obj)
+        return {};
+
+    addr_t file = obj + this->offsets[OBJECT_HEADER_BODY];
+    addr_t filename = file + this->offsets[FILE_OBJECT_FILENAME];
+    addr_t filetype = file + this->offsets[FILE_OBJECT_TYPE];
+
+    if (out_file)
+        *out_file = file;
+
+    if (out_filetype)
+        *out_filetype = filetype;
+
+    ACCESS_CONTEXT(ctx);
+    ctx.translate_mechanism = VMI_TM_PROCESS_DTB;
+    ctx.addr = filetype;
+    ctx.dtb = info->regs->cr3;
+
+    uint8_t type = 0;
+    if (VMI_FAILURE == vmi_read_8(vmi, &ctx, &type))
+        return {};
+
+    if (type != 5)
+        return {};
+
+    unicode_string_t* filename_us = drakvuf_read_unicode(drakvuf,
+            info,
+            filename);
+    if (!filename_us) return {};
+    std::string ret = {(const char*)filename_us->contents};
+    vmi_free_unicode_str(filename_us);
+    return ret;
+}
+
+void fileextractor::print_filedelete_information(drakvuf_trap_info_t* info,
+    task_t& task)
+{
+    std::string flags = parse_flags(task.fo_flags, fo_flags_map, this->format);
+    std::string r;
+    switch (task.reason)
+    {
+        case task_t::task_reason::write:
+            r = "WriteFile";
+            break;
+        case task_t::task_reason::del:
+            r = "DeleteFile";
+            break;
+        default:
+            r = "Unknown";
+            break;
+    }
+
+    if (this->format == OUTPUT_KV)
+    {
+        kvfmt::print("fileextractor", drakvuf, info,
+            keyval("FileName", fmt::Qstr(task.filename)),
+            keyval("Size", fmt::Nval(task.file_size)),
+            keyval("Flags", fmt::Xval(task.fo_flags)),
+            fmt::Rstr(flags),
+            keyval("SeqNum", fmt::Nval(task.idx)),
+            keyval("Reason", fmt::Qstr(r))
+        );
+    }
+    else
+    {
+        fmt::print(this->format, "fileextractor", drakvuf, info,
+            keyval("FileName", fmt::Qstr(task.filename)),
+            keyval("Size", fmt::Nval(task.file_size)),
+            keyval("Flags", fmt::Xval(task.fo_flags)),
+            keyval("FlagsExpanded", fmt::Qstr(flags)),
+            keyval("SeqNum", fmt::Nval(task.idx)),
+            keyval("Reason", fmt::Qstr(r))
+        );
+    }
+}
+
+void fileextractor::print_extraction_failure(drakvuf_trap_info_t* info,
+    const string& filename,
+    const string& message)
+{
+    fmt::print(this->format, "fileextractor_fail", drakvuf, info,
+        keyval("FileName", fmt::Qstr(filename)),
+        keyval("Message", fmt::Qstr(message))
+    );
+}
+
+void fileextractor::save_file_metadata(drakvuf_trap_info_t* info,
+    addr_t control_area,
+    task_t& task)
+{
+    char* file = NULL;
+    if ( asprintf(&file,
+            "%s/file.%06d.metadata",
+            this->dump_folder,
+            task.idx) < 0 )
+        return;
+
+    FILE* fp = fopen(file, "w");
+    free(file);
+    if (!fp)
+        return;
+
+    json_object* jobj = json_object_new_object();
```

Comment:
maybe use fmt instead of raw json?

Reply:
Why should we do this?

The usage of `fmt` would be against the switch to json-c library by @icedevml from https://github.com/tklengyel/drakvuf/commit/048e9e343a800ecabaeeb8f958c6341c1b837f97